File: Tree/Binary Search Tree/go/src/bst.go
package bst

import (
	"errors"
)

type Node struct {
	Value string
	Data  string
	Left  *Node
	Right *Node
}

// insert into node
func (n *Node) Insert(value, data string) error {
	if n == nil {
		return errors.New("Cannot insert a value into a nil tree")
	}

	switch {
	case value == n.Value:
		return nil
	case value < n.Value:
		if n.Left == nil {
			n.Left = &Node{Value: value, Data: data}
			return nil

		}
		return n.Left.Insert(value, data)

	case value < n.Value:
		if n.Right == nil {
			n.Right = &Node{Value: value, Data: data}
			return nil
		}

		return n.Right.Insert(value, data)
	}
	return nil
}

// find element
func (n *Node) Find(s string) (string, bool) {
	if n == nil {
		return "", false
	}

	switch {
	case s == n.Value:
		return n.Data, true
	case s < n.Value:
		return n.Left.Find(s)
	default:
		return n.Right.Find(s)
	}
}

func (n *Node) findMax(parent *Node) (*Node, *Node) {
	if n == nil {
		return nil, parent
	}

	if n.Right == nil {
		return n, parent
	}

	return n.Right.findMax(n)
}

func (n *Node) replaceNode(parent, replacement *Node) error {
	if n == nil {
		return errors.New("replaceNode() not allowed on a nil node")
	}

	if n == parent.Left {
		parent.Left = replacement
		return nil
	}

	parent.Right = replacement
	return nil
}

func (n *Node) Delete(s string, parent *Node) error {
	if n == nil {
		return errors.New("Value to be deleted does not exist in the tree")
	}

	switch {
	case s < n.Value:
		return n.Left.Delete(s, n)
	case s > n.Value:
		return n.Right.Delete(s, n)
	default:
		if n.Left == nil && n.Right == nil {
			n.replaceNode(parent, nil)
			return nil
		}

		if n.Left == nil {
			n.replaceNode(parent, n.Right)
			return nil
		}

		if n.Right == nil {
			n.replaceNode(parent, n.Left)
			return nil
		}

		replacement, replParent := n.Left.findMax(n)

		n.Value = replacement.Value
		n.Data = replacement.Data

		return replacement.Delete(replacement.Value, replParent)
	}
}

type Tree struct {
	Root *Node
}

func (t *Tree) Insert(value, data string) error {
	if t.Root == nil {
		t.Root = &Node{Value: value, Data: data}
		return nil
	}

	return t.Root.Insert(value, data)
}

func (t *Tree) Find(s string) (string, bool) {
	if t.Root == nil {
		return "", false
	}

	return t.Root.Find(s)
}

func (t *Tree) Delete(s string) error {
	if t.Root == nil {
		return errors.New("Cannot delete from an empty tree")
	}

	fakeParent := &Node{Right: t.Root}
	err := t.Root.Delete(s, fakeParent)
	if err != nil {
		return err
	}

	if fakeParent.Right == nil {
		t.Root = nil
	}
	return nil
}

func (t *Tree) Traverse(n *Node, f func(*Node)) {
	if n == nil {
		return
	}
	t.Traverse(n.Left, f)
	f(n)
	t.Traverse(n.Right, f)
}

// Thank you to https://appliedgo.net/bintree/

File: Other Algorithms/BST.cpp
#include <iostream>

using namespace std;

class node{

  public:

  int val;
  int element;
  node* left;
  node* right;

  node(int val,node*left,node*right){
    this->val=val;
    this->left=left;
    this->right=right;
  }

};


class bst{

  public:

  node* root;
  int num;

  bst(){
    num=0;
    root=NULL;
  }

  node* insertr(node* no,int k){
    if(no==NULL) {
      no=new node(k,NULL,NULL);
      return no;
    }
    if(k<no->val){
      no->left=insertr(no->left,k);
      return no;
    }
    else{
      no->right=insertr(no->right,k);
      return no;
    }

  }

  void insert(int k){
    if(this->root==NULL) {
      this->root=new node(k,NULL,NULL);
      num++;
      exit(0);
    }
    if(k<this->root->val){
      this->root->left=insertr(root->left,k);
      num++;
      exit(0);
    }
    else{
      this->root->right=insertr(root->right,k);
      num++;
      exit(0);
    }
    exit(-1);
  }

  bool findr(node*no,int k){
    if(no==NULL) return false;
    if(k==this->root->val) return true;
    if(k<no->val){
        return findr(no->left,k);
      }
      else{
        return findr(no->right,k);
      }

  }

  bool find(int k){
    if(num>0){
      if(k==this->root->val) return true;

      if(k<this->root->val){
        return findr(this->root->left,k);
      }
      else{
        return findr(this->root->right,k);
      }
    }
    return false;
  }
  

};

int main() {
  string str;
  cin>>str;
  int*nha=NULL;
  nha[55]=3;
}
